File: FilePicker/js/js/scenario3_SingleFolder.js
//// THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY OF
//// ANY KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO
//// THE IMPLIED WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A
//// PARTICULAR PURPOSE.
////
//// Copyright (c) Microsoft Corporation. All rights reserved

(function () {
    "use strict";
    var page = WinJS.UI.Pages.define("/html/scenario3_SingleFolder.html", {
        ready: function (element, options) {
            element.querySelector(".folder").addEventListener("click", pickFolder, false);
        }
    });

    function pickFolder() {
        // Clean scenario output
        WinJS.log && WinJS.log("", "sample", "status");

        // Create the picker object and set options
        var folderPicker = new Windows.Storage.Pickers.FolderPicker;
        folderPicker.suggestedStartLocation = Windows.Storage.Pickers.PickerLocationId.desktop;
        // Users expect to have a filtered view of their folders depending on the scenario.
        // For example, when choosing a documents folder, restrict the filetypes to documents for your application.
        folderPicker.fileTypeFilter.replaceAll([".docx", ".xlsx", ".pptx"]);

        folderPicker.pickSingleFolderAsync().then(function (folder) {
            if (folder) {
                // Application now has read/write access to all contents in the picked folder (including sub-folder contents)
                // Cache folder so the contents can be accessed at a later time
                Windows.Storage.AccessCache.StorageApplicationPermissions.futureAccessList.addOrReplace("PickedFolderToken", folder);
                WinJS.log && WinJS.log("Picked folder: " + folder.name, "sample", "status");
            } else {
                // The picker was dismissed with no selected file
                WinJS.log && WinJS.log("Operation cancelled.", "sample", "status");
            }
        });
    }
})();

File: Samples/FilePicker/cppwinrt/Scenario3_SingleFolder.cpp
//*********************************************************
//
// Copyright (c) Microsoft. All rights reserved.
// This code is licensed under the MIT License (MIT).
// THIS CODE IS PROVIDED *AS IS* WITHOUT WARRANTY OF
// ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING ANY
// IMPLIED WARRANTIES OF FITNESS FOR A PARTICULAR
// PURPOSE, MERCHANTABILITY, OR NON-INFRINGEMENT.
//
//*********************************************************

#include "pch.h"
#include "Scenario3_SingleFolder.h"
#include "Scenario3_SingleFolder.g.cpp"

using namespace winrt;
using namespace Windows::Foundation;
using namespace Windows::Storage;
using namespace Windows::Storage::AccessCache;
using namespace Windows::Storage::Pickers;
using namespace Windows::UI::Xaml;

namespace winrt::SDKTemplate::implementation
{
    Scenario3_SingleFolder::Scenario3_SingleFolder()
    {
        InitializeComponent();
    }

    fire_and_forget Scenario3_SingleFolder::PickFolderButton_Click(IInspectable const&, RoutedEventArgs const&)
    {
        auto lifetime = get_strong();

        // Clear previous returned folder name, if it exists, between iterations of this scenario
        OutputTextBlock().Text(L"");

        FolderPicker folderPicker;
        folderPicker.SuggestedStartLocation(PickerLocationId::Desktop);
        folderPicker.FileTypeFilter().ReplaceAll({ L".docx",  L".xlsx", L".pptx" });
        StorageFolder folder = co_await folderPicker.PickSingleFolderAsync();
        if (folder != nullptr)
        {
            // The StorageFolder has read/write access to all contents in the picked folder (including other sub-folder contents).
            // See the FileAccess sample for code that obtains a StorageFile from a StorageFolder to read and write.
            StorageApplicationPermissions::FutureAccessList().AddOrReplace(L"PickedFolderToken", folder);
            OutputTextBlock().Text(L"Picked folder: " + folder.Name());
        }
        else
        {
            OutputTextBlock().Text(L"Operation cancelled.");
        }
    }

}
